### PR TITLE
[Notification] Handlers + recipients

### PR DESCRIPTION
**Pull request by qnkhuat**
The 3rd bullet point of milestone 1 in https://github.com/metabase/metabase/issues/47438

This introduces 2 tables : notification_handler and notification_recipient and utility function to create them

### CODE REVIEW

File: resources/migrations/001_update_migrations.yaml [hunk 1]
```diff
@@ -9281,7 +9281,7 @@ databaseChangeLog:
       changes:
         - createTable:
             tableName: notification
-            remarks: join table that connect notification subscriptions and notification destinations
+            remarks: join table that connect notification subscriptions and notification handlers
```
Comment (crisptrutski) [inline resources/migrations/001_update_migrations.yaml:9284]: I guess its fine to leave this rename here, and keeping as "destinations" in the previous PR.

File: resources/migrations/001_update_migrations.yaml [hunk 2]
```diff
@@ -9383,6 +9383,242 @@ databaseChangeLog:
...
+            remarks: which channel to send the notification to
+            columns:
+              - column:
+                  name: id
+                  type: int
+                  autoIncrement: true
+                  constraints:
+                    primaryKey: true
+                    nullable: false
+              # TODO: can be removed once emails and slack configuration are fully migrated to be stored in channel table.
```
Comment (crisptrutski) [inline resources/migrations/001_update_migrations.yaml:9406]: This will be nice

File: resources/migrations/001_update_migrations.yaml [hunk 3]
```diff
@@ -9383,6 +9383,242 @@ databaseChangeLog:
...
+              - column:
+                  name: channel_type
+                  type: varchar(64)
+                  remarks: the type of the channel
```
Comment (crisptrutski) [inline resources/migrations/001_update_migrations.yaml:9410]: Worth mentioning the concrete values here?

File: resources/migrations/001_update_migrations.yaml [hunk 4]
```diff
@@ -9383,6 +9383,242 @@ databaseChangeLog:
...
+              - column:
+                  name: type
+                  type: varchar(64)
+                  remarks: the type of the recipient
+                  constraints:
+                    nullable: false
+              - column:
+                  name: user_id
+                  type: int
+                  remarks: an user if the recipient has type user
```
Comment (crisptrutski) [inline resources/migrations/001_update_migrations.yaml:9530]: [English is weird](https://english.stackexchange.com/a/105117) (article is based on sound, not spelling)

```suggestion
                  remarks: a user if the recipient has type user
```

File: resources/migrations/001_update_migrations.yaml [hunk 5]
```diff
@@ -9383,6 +9383,242 @@ databaseChangeLog:
...
+            tableName: notification_handler
+            columns:
+              - column:
+                  name: channel_id
+            indexName: idx_notification_handler_channel_id
+
+  - changeSet:
+      id: v51.2024-09-05T08:00:07
+      author: qnkhuat
+      comment: Create the notification_recipient table
```
Comment (crisptrutski) [inline resources/migrations/001_update_migrations.yaml:9494]: Will we not need to support external recipients? (i.e. a plain email address)

File: src/metabase/models/notification.clj [hunk 6]
```diff
@@ -54,12 +121,19 @@
                           [:= :ns.event_name (u/qualified-name event-name)]
                           [:= :ns.type (u/qualified-name :notification-subscription/system-event)]]}))
 
-(defn create-notification!
+(mu/defn create-notification!
```
Comment (crisptrutski) [inline src/metabase/models/notification.clj:124]: You don't seem to have added any type annotations to go with this

File: resources/migrations/001_update_migrations.yaml (hunk 1 above)
Comment (crisptrutski) [reply resources/migrations/001_update_migrations.yaml:9284]: Hmmm, actually I wonder if this is what's breaking the checksum in the rollback tests. Silly for the comment to be in the hash 😑

**Review by crisptrutski — APPROVED** [review]
This modelling seems quite natural. Had a few questions and non-functional suggestions.

File: resources/migrations/001_update_migrations.yaml (hunk 1 above)
Comment (qnkhuat) [reply resources/migrations/001_update_migrations.yaml:9284]: Yeah, it's what is breaking the checksum!

File: resources/migrations/001_update_migrations.yaml (hunk 4 above)
Comment (qnkhuat) [reply resources/migrations/001_update_migrations.yaml:9612]: I kept forgetting this!

File: resources/migrations/001_update_migrations.yaml (hunk 5 above)
Comment (qnkhuat) [reply resources/migrations/001_update_migrations.yaml:9576]: yes, in that case we'll have 
```clojure
{:type :notification-recipient/external-email
 :details "chris@metabase.com"}
```

`details` is different for each type and used for custom ones

File: src/metabase/models/notification.clj (hunk 6 above)
Comment (qnkhuat) [reply src/metabase/models/notification.clj]: yea, I'm adding one in the next pr, but will remove that mu/ for now

File: resources/migrations/001_update_migrations.yaml (hunk 3 above)
Comment (qnkhuat) [reply resources/migrations/001_update_migrations.yaml]: added

File: resources/migrations/001_update_migrations.yaml (hunk 5 above)
Comment (crisptrutski) [reply resources/migrations/001_update_migrations.yaml:9576]: Ah ok. What do you think about `:configuration` instead? Details sounds non-functional.

Comment (qnkhuat) [reply resources/migrations/001_update_migrations.yaml:9576]: yea configuration does sound better, I'll update it in the upstream PR

Comment (qnkhuat) [reply resources/migrations/001_update_migrations.yaml:9576]: on JSON, yes!

**Pull request merged**